### ISSUE TITLE
unit performance benchmark

### ISSUE THREAD

**Issue by ngoldbaum** [astropy/astropy#7546]
I just submitted a paper to the Journal of Open Source Software presenting [unyt](github.com/yt-project/unyt), a new Python library I'm releasing based on `yt.units`. The idea is that it will enable people to use yt's unit system without depending on all of yt. Eventually we will make `yt.units` use `unyt` under the hood.

The reason I'm opening an issue is that part of the paper includes a performance benchmark (see benchmarks/bench.py in the unyt repo) comparing the performance of `unyt`, `Pint`, and `astropy.units`. In general `astropy.units` does pretty well, generally coming in second place behind `unyt`, except for the benchmarks of the `sqrt` and `power` function, where it comes in last place.

You can see a preprint of the paper in the unyt repository:

https://github.com/yt-project/unyt/blob/master/paper/paper.pdf

It will also hopefully be on the arxiv soon (assuming it makes it past the arxiv moderators). In principle the library will be getting updates from the JOSS review process, but I'm expecting the results to not change much before the paper is accepted.

I'm opening this issue not as a challenge or as a brag, but in the hope that all three unit libraries (and their users) will benefit from the benchmarks. I'm also very happy to help out with reproducing the benchmarks. I know that performance is a major focus for the next major release of astropy, so hopefully this is helpful.

**Comment by astrofrog** [comment]
Nice to see that the yt unit framework has been split out into a separate package! This is an idea we've toyed with in the past for astropy (and haven't definitely ruled out). Is the code for the benchmarks available somewhere?

Just a small comment on the paper:

*This means that depending on astropy.units requires depending on a large collection of astronomically focused software, including a substantial amount of compiled C code. This presents a barrier to usage for potential users of astropy.units who are not astronomers or do not need the observational astronomy capabilities provided by astropy*

I feel this is not quite fair - depending on astropy.units doesn't add additional dependencies (beyond numpy). The package is larger to install and includes C code, but the current phrasing makes it sound like additional package dependencies are required. Of course it's still fine to say that unyt is much easier to install, I'm just suggesting a more careful use of the term 'dependency' above.

**Comment by ngoldbaum** [comment]
I've updated the text about astropy, see https://github.com/yt-project/unyt/commit/ef3338de5b9429ae9af17f31373ba0d50f006b09.

The benchmark is located at `benchmarks/bench.py`, see: https://github.com/yt-project/unyt/blob/master/benchmarks/bench.py. The code to generate the benchmark plots is in `paper/benchmark_plot.py`, see https://github.com/yt-project/unyt/blob/master/paper/benchmark_plot.py.

The benchmarks use Victor Stinner's `perf` tool. To get the most out of `perf` you need a computer with linux installed and root access. It needs to be linux running on actual bare metal, not a VM, so that `perf` can adjust linux kernel parameters.

**Comment by astrofrog** [comment]
Note to astropy developers: we should add these specific benchmarks to astropy-benchmarks so that we can work on improving these kinds of operations.

**Comment by mhvk** [comment]
@ngoldbaum - thanks for the link! Looks like we need to see more carefully why `unit ** 0.5` is so slow...

Some comments: I do not see how a unit library can ever be faster than numpy, yet that is what you find for multiplication and division - this suggests you do some explicit unit conversion for numpy even though it is not necessary. But... maybe that is in fact reasonable: if one writes code with straight numbers, one would have to ensure the implicit units are actually correct, while with a units package that only needs to be done when the numbers are needed.

**Comment by ngoldbaum** [comment]
That’s basically it yeah. This is described in the text a bit but I guess I
should probably add a short note in the figure caption.

You can see what's happening looking at these snippets:

In [6]: import astropy.units as u
   ...: data1 = [1, 2, 3]*u.km
   ...: data2 = [1, 2, 3]*u.m
   ...: data1/data2
   ...:
Out[6]: <Quantity [1., 1., 1.] km / m>

In [12]: import unyt as u
    ...: data1 = [1, 2, 3]*u.km
    ...: data2 = [1, 2, 3]*u.m
    ...: data1/data2
    ...:
Out[12]: unyt_array([1000., 1000., 1000.], '(dimensionless)')

With arrays of a million elements it matters a lot if you need to multiply
by 1000 or not. On the other hand, units like km/m are confusing to users,
so I actually like this feature in unyt. Both Pint and astropy.units don't
reduce the intermediate expression.

**Comment by ngoldbaum** [comment]
That’s basically it yeah. This is described in the text a bit but I guess I
should probably add a short note in the figure caption.

You can see what's happening looking at these snippets:

```python
>>> import astropy.units as u
>>> data1 = [1, 2, 3]*u.km
>>> data2 = [1, 2, 3]*u.m
>>> data1/data2
<Quantity [1., 1., 1.] km / m>

>>> import unyt as u
>>> data1 = [1, 2, 3]*u.km
>>> data2 = [1, 2, 3]*u.m
>>> data1/data2
unyt_array([1000., 1000., 1000.], '(dimensionless)')
```

With arrays of a million elements it matters a lot if you need to multiply
by 1000 or not. On the other hand, units like km/m are confusing to users,
so I actually like this feature in unyt. Both Pint and astropy.units don't
reduce the intermediate expression.

**Comment by mhvk** [comment]
Yes, we do not reduce units on purpose, since it is not all that unlikely that next the result will be multiplied by something in `m`...

**Comment by pllim** [comment]
Benchmarks are added in  astropy/astropy-benchmarks#54 . Where do we go from here? Close?

**Comment by astrofrog** [comment]
I think the questions is whether there are specific cases that we still need to improve the performance for for 3.1 - if so, I think we should open dedicated issues and close this one.

**Comment by pllim** [comment]
> we should open dedicated issues and close this one

Sounds good. @mhvk ?

**Comment by mhvk** [comment]
Agreed. I don't think there were any specific items that should out except for taking powers (addressed by #7549). So, closing. Again thanks for sharing, @ngoldbaum!
